File: src/cse/CSEMachine.java
package cse;
import Interpreter.OperationHandler;
import cse.ele.EleTuple;
import cse.ele.EleValue;
import cse.ele.EleValueOrTuple;

import java.util.ArrayList;

/**
 * CSE machine
 */

public class CSEMachine {
    private final Stack<EleValue> eleValues;
    private final Stack<EleValueOrTuple> eleValueOrTuples;
    private final OperationHandler operationHandler;
    private final ArrayList<Environment> environments;
    private final ArrayList<Stack<EleValue>> CS;

    public CSEMachine(ArrayList<Stack<EleValue>> controlStructures) {
        this.CS = controlStructures;
        this.eleValueOrTuples = new Stack<>();
        this.operationHandler = new OperationHandler();

        eleValues = new Stack<>();
        eleValues.push(new EleValue("environment", "0"));
        eleValues.push(new EleValue("delta", "0"));
        eleValueOrTuples.push(new EleValue("environment", "0"));
        environments = new ArrayList<>();
        environments.add(new Environment());
    }

    @Override
    public String toString() {
        return eleValues + "\n" + eleValueOrTuples + "\n" + currentEnvironment() + "\n";
    }

    /**
     * By peering at the stack and locating the nearest environment element, one can find the current environment index.
     */
    private int currentEnvironmentIndex() {
        int closestEnvironment = 0;
        for (EleValueOrTuple element : eleValues) {
            if (element instanceof EleValue && element.isLabel("environment")) {
                String closestEnvironmentStr = ((EleValue) element).getValue();
                closestEnvironment = Integer.parseInt(closestEnvironmentStr);
            }
        }
        return closestEnvironment;
    }

    /**
     * Get current environment
     */
    private Environment currentEnvironment() {
        return environments.get(currentEnvironmentIndex());
    }

    /**
     * Start analyzing the control stack to determine the outcome.
     */
    public void evaluateTree() {
        while (!eleValues.isEmpty()) {
            EleValue currentElement = eleValues.pop();

            if (currentElement.isLabel("gamma")) {
                EleValueOrTuple firstElem = eleValueOrTuples.pop();
                EleValueOrTuple secondElem = eleValueOrTuples.pop();
                if (firstElem.isLabel("yStar")) {
                    Rule12(secondElem);
                } else if (firstElem.isLabel("eta")) {
                    eleValueOrTuples.push(secondElem);
                    Rule13(currentElement, firstElem);
                } else if (firstElem.isLabel("lambda")) {
                    EleValue firstValue = (EleValue) firstElem;
                    if (firstValue.getValue().contains(",")) {
                        Rule11(firstElem, secondElem);
                    } else {
                        Rule4(firstElem, secondElem);
                    }
                } else if (firstElem.isLabel("tau")) {
                    Rule10(firstElem, secondElem);
                } else {
                    Rule3(firstElem, secondElem);
                }
            } else if (currentElement.isLabel("delta")) {
                int controlIndex = Integer.parseInt(currentElement.getValue());
                extractDelta(controlIndex);
            } else if (currentElement.isLabel("id")) {
                Rule1(currentElement);
            } else if (currentElement.isLabel("lambda")) {
                Rule2(currentElement);
            } else if (currentElement.isLabel("environment")) {
                Rule5(currentElement);
            } else if (currentElement.isLabel("beta")) {
                Rule8();
            } else if (currentElement.isLabel("tau")) {
                Rule9(currentElement);
            } else if (!Rule6_7(currentElement)) {
                eleValueOrTuples.push(currentElement);
            }
        }
    }

    /**
     * Extract components from the specified control structure.
     */
    private void extractDelta(int controlIndex) {
        Stack<EleValue> control = CS.get(controlIndex);
        for (EleValue controlElem : control) {
            this.eleValues.push(controlElem);
        }
    }

    /**
     * rule 1
     */
    private void Rule1(EleValue name) {
        String id = name.getValue();
        EleValueOrTuple value = currentEnvironment().lookup(id);
        if (value == null) {
            value = new EleValue(id);
        }
        eleValueOrTuples.push(value);
    }

    /**
     * rule 2
     */
    private void Rule2(EleValue lambda) {
        String[] kAndX = lambda.getValue().split(" ");
        String c = Integer.toString(currentEnvironmentIndex());
        String[] newValues = {kAndX[0], kAndX[1], c};
        EleValueOrTuple newLambda = new EleValue("lambda", String.join(" ", newValues));
        eleValueOrTuples.push(newLambda);
    }

    /**
     * rule3
     */
    private void Rule3(EleValueOrTuple rator, EleValueOrTuple rand) {
        EleValueOrTuple result = operationHandler.apply(rator, rand);
        eleValueOrTuples.push(result);
    }

    /**
     * rule4
     */
    private void Rule4(EleValueOrTuple lambda, EleValueOrTuple rand) {
        if (lambda instanceof EleValue && lambda.isLabel("lambda")) {
            String[] kAndXAndC = ((EleValue) lambda).getValue().split(" ");
            String k = kAndXAndC[0];
            String x = kAndXAndC[1];
            String c = kAndXAndC[2];
            Environment envC = environments.get(Integer.parseInt(c));

            Environment newEnvironment = new Environment(envC, x, rand);
            String newEnvIndex = Integer.toString(environments.size());
            environments.add(newEnvironment);
            eleValues.push(new EleValue("environment", newEnvIndex));
            eleValues.push(new EleValue("delta", k));
            eleValueOrTuples.push(new EleValue("environment", newEnvIndex));
            return;
        }
        throw new ExceptionHandlerOfCSE("Expected lambda element but found: " + lambda);
    }


    /**
     * rule 5
     */
    private void Rule5(EleValue env) {
        EleValueOrTuple value = eleValueOrTuples.pop();
        EleValueOrTuple envS = eleValueOrTuples.pop();
        if (envS instanceof EleValue && envS.isLabel("environment")) {
            if (env.equals(envS)) {
                eleValueOrTuples.push(value);
                return;
            }
            throw new ExceptionHandlerOfCSE(String.format("Environment element mismatch: %s and %s", env, envS));
        }
        throw new ExceptionHandlerOfCSE("Expected environment element but found: " + envS);
    }

    /**
     * rule7
     */
    private boolean Rule6_7(EleValue element) {
        if (operationHandler.checkMathematicalOperation(element)) {
            EleValueOrTuple rator = eleValueOrTuples.pop();
            EleValueOrTuple rand = eleValueOrTuples.pop();
            EleValueOrTuple result = operationHandler.applyOperations(element, rator, rand);
            eleValueOrTuples.push(result);
        } else if (operationHandler.checkArrayOperation(element)) {
            EleValueOrTuple rand = eleValueOrTuples.pop();
            EleValueOrTuple result = operationHandler.apply(element, rand);
            eleValueOrTuples.push(result);
        } else {
            return false;
        }
        return true;
    }

    /**
     * rule8
     */
    private void Rule8() {
        EleValue deltaElse = eleValues.pop();
        EleValue deltaThen = eleValues.pop();
        EleValueOrTuple condition = eleValueOrTuples.pop();

        if (deltaElse.isLabel("delta") && deltaThen.isLabel("delta")) {
            if (condition.isLabel("true")) {
                eleValues.push(deltaThen);
                return;
            } else if (condition.isLabel("false")) {
                eleValues.push(deltaElse);
                return;
            }
            throw new RuntimeException("If condition must evaluate to a truth value.");
        }
        throw new ExceptionHandlerOfCSE("Expected delta elements.");
    }

    /**
     * rule 9
     */
    private void Rule9(EleValue tau) {
        int elements = Integer.parseInt(tau.getValue());
        EleValueOrTuple[] tupleElements = new EleValueOrTuple[elements];
        for (int i = 0; i < elements; i++) {
            tupleElements[i] = eleValueOrTuples.pop();
        }
        EleValueOrTuple tuple = new EleTuple(tupleElements);
        eleValueOrTuples.push(tuple);
    }

    /**
     * rule10
     */
    private void Rule10(EleValueOrTuple tuple, EleValueOrTuple index) {
        if (tuple instanceof EleTuple) {
            if (index instanceof EleValue && index.isLabel("int")) {
                int ind = Integer.parseInt(((EleValue) index).getValue());
                EleValueOrTuple value = ((EleTuple) tuple).getValue()[ind];
                eleValueOrTuples.push(value);
                return;
            }
            throw new ExceptionHandlerOfCSE("Expected integer index but found: " + index);
        }
        throw new ExceptionHandlerOfCSE("Expected tuple but found: " + tuple);
    }

    /**
     * rule 11
     */
    private void Rule11(EleValueOrTuple lambda, EleValueOrTuple rand) {
        if (lambda instanceof EleValue && lambda.isLabel("lambda")) {
            if (rand instanceof EleTuple) {
                String[] kAndVAndC = ((EleValue) lambda).getValue().split(" ");
                String k = kAndVAndC[0];
                String[] v = kAndVAndC[1].split(",");
                String c = kAndVAndC[2];
                Environment envC = environments.get(Integer.parseInt(c));

                Environment newEnvironment = new Environment(envC);
                for (int i = 0; i < v.length; i++) {
                    newEnvironment.remember(v[i], ((EleTuple) rand).getValue()[i]);
                }
                String newEnvIndex = Integer.toString(environments.size());
                environments.add(newEnvironment);
                eleValues.push(new EleValue("environment", newEnvIndex));
                eleValues.push(new EleValue("delta", k));
                eleValueOrTuples.push(new EleValue("environment", newEnvIndex));
                return;
            }
            throw new ExceptionHandlerOfCSE("Expected tuple but found: " + rand);
        }
        throw new ExceptionHandlerOfCSE("Expected lambda element but found: " + lambda);
    }

    /**
     * rule12
     */
    private void Rule12(EleValueOrTuple lambda) {
        if (lambda instanceof EleValue && lambda.isLabel("lambda")) {
            String iAndVAndC = ((EleValue) lambda).getValue();
            EleValueOrTuple etaElement = new EleValue("eta", iAndVAndC);
            eleValueOrTuples.push(etaElement);
            return;
        }
        throw new ExceptionHandlerOfCSE("Expected lambda element but found: " + lambda);
    }

    /**
     * rule 13
     */
    private void Rule13(EleValue gamma, EleValueOrTuple eta) {
        if (eta instanceof EleValue && eta.isLabel("eta")) {
            String iAndVAndC = ((EleValue) eta).getValue();
            EleValue lambda = new EleValue("lambda", iAndVAndC);
            EleValue newGamma = new EleValue("gamma");

            eleValueOrTuples.push(eta);
            eleValueOrTuples.push(lambda);

            eleValues.push(gamma);
            eleValues.push(newGamma);
            return;
        }
        throw new ExceptionHandlerOfCSE("Expected eta element but found: " + eta);
    }
}

File: reCreate/main/CSE/CSEMachine.py
from .elements.EleValue import EleValue
from .elements.EleTuple import EleTuple
from .elements.EleValueOrTuple import EleValueOrTuple
from .ExceptionHandlerOfCSE import ExceptionHandlerOfCSE
from INTERPRETER.OperationHandler import OperationHandler
from .Environment import Environment
from collections import deque

class CSEMachine:
    def __init__(self, controlStructures):
        self.CS = controlStructures
        self.eleValueOrTuples = deque()
        self.operationHandler = OperationHandler()

        self.eleValues = deque()
        self.eleValues.append(EleValue("environment", "0"))
        self.eleValues.append(EleValue("delta", "0"))
        self.eleValueOrTuples.append(EleValue("environment", "0"))
        self.environments = [Environment()]

    def toString(self):
        return str(self.eleValues) + "\n" + str(self.eleValueOrTuples) + "\n" + str(self.currentEnvironment()) + "\n"

    def currentEnvironmentIndex(self):
        closestEnvironment = 0
        for element in self.eleValues:
            if isinstance(element, EleValue) and element.isLabel("environment"):
                closestEnvironmentStr = element.getValue()
                closestEnvironment = int(closestEnvironmentStr)
        return closestEnvironment

    def currentEnvironment(self):
        return self.environments[self.currentEnvironmentIndex()]

    def evaluateTree(self):
        counter=0
      
        while self.eleValues:
            counter+=1
            # print(counter)
            currentElement = self.eleValues.pop()
            # print(currentElement.getLabel())
            if currentElement.isLabel("gamma"):
                firstElem = self.eleValueOrTuples.pop()
                secondElem = self.eleValueOrTuples.pop()
                if firstElem.isLabel("yStar"):
                    self.rule12(secondElem)
                elif firstElem.isLabel("eta"):
                    self.eleValueOrTuples.append(secondElem)
                    self.rule13(currentElement, firstElem)
                elif firstElem.isLabel("lambda"):
                    firstValue = firstElem
                    if "," in firstValue.getValue():
                        self.rule11(firstElem, secondElem)
                    else:
                        self.rule4(firstElem, secondElem)
                elif firstElem.isLabel("tau"):
                    self.rule10(firstElem, secondElem)
                else:
                   
                    self.rule3(firstElem, secondElem)
            elif currentElement.isLabel("delta"):
                controlIndex = int(currentElement.getValue())
                self.extractDelta(controlIndex)
            elif currentElement.isLabel("id"):
                self.rule1(currentElement)
            elif currentElement.isLabel("lambda"):
                self.rule2(currentElement)
            elif currentElement.isLabel("environment"):
                self.rule5(currentElement)
            elif currentElement.isLabel("beta"):
                self.rule8()
            elif currentElement.isLabel("tau"):
                self.rule9(currentElement)
            elif not self.rule6_7(currentElement):
                self.eleValueOrTuples.append(currentElement)
        
    def extractDelta(self, controlIndex):
        control = self.CS[controlIndex]
        for controlElem in control:
            self.eleValues.append(controlElem)

    def rule1(self, name):
        id = name.getValue()
        value = self.currentEnvironment().lookup(id)
        if value is None:
            value = EleValue(id)
        self.eleValueOrTuples.append(value)

    def rule2(self, lambdaVal):
        kAndX = lambdaVal.getValue().split(" ")
        c = str(self.currentEnvironmentIndex())
        newValues = [kAndX[0], kAndX[1], c]
        newLambda = EleValue("lambda", " ".join(newValues))
        self.eleValueOrTuples.append(newLambda)

    def rule3(self, rator, rand):
        
        result = self.operationHandler.apply(rator, rand)
        self.eleValueOrTuples.append(result)

    def rule4(self, lambdaVal, rand):
        if isinstance(lambdaVal, EleValue) and lambdaVal.isLabel("lambda"):
            kAndXAndC = lambdaVal.getValue().split(" ")
            k, x, c = kAndXAndC[0], kAndXAndC[1], kAndXAndC[2]
            envC = self.environments[int(c)]

            newEnvironment = Environment(envC, x, rand)
            newEnvIndex = str(len(self.environments))
            self.environments.append(newEnvironment)
            self.eleValues.append(EleValue("environment", newEnvIndex))
            self.eleValues.append(EleValue("delta", k))
            self.eleValueOrTuples.append(EleValue("environment", newEnvIndex))
            return
        raise ExceptionHandlerOfCSE("Expected lambda element but found: " + lambdaVal)

    def rule5(self, env):
       
        value = self.eleValueOrTuples.pop()
        envS = self.eleValueOrTuples.pop()
        if isinstance(envS, EleValue) and envS.isLabel("environment"):
          
            if env.equals(envS):
                self.eleValueOrTuples.append(value)
                return
            raise ExceptionHandlerOfCSE("Environment element mismatch: {} and {}".format(env, envS))
        raise ExceptionHandlerOfCSE("Expected environment element but found: " + envS)

    def rule6_7(self, element):
        if self.operationHandler.checkMathematicalOperation(element):
            rator = self.eleValueOrTuples.pop()
            rand = self.eleValueOrTuples.pop()
            result = self.operationHandler.applyOperations(element, rator, rand)
            self.eleValueOrTuples.append(result)
        elif self.operationHandler.checkArrayOperation(element):
            rand = self.eleValueOrTuples.pop()
            result = self.operationHandler.apply(element, rand)
            self.eleValueOrTuples.append(result)
        else:
            return False
        return True

    def rule8(self):
        deltaElse = self.eleValues.pop()
        deltaThen = self.eleValues.pop()
        condition = self.eleValueOrTuples.pop()

        if deltaElse.isLabel("delta") and deltaThen.isLabel("delta"):
            if condition.isLabel("true"):
                self.eleValues.append(deltaThen)
                return
            elif condition.isLabel("false"):
                self.eleValues.append(deltaElse)
                return
            raise RuntimeError("If condition must evaluate to a truth value.")
        raise ExceptionHandlerOfCSE("Expected delta elements.")

    def rule9(self, tau):
        elements = int(tau.getValue())
        tupleElements = [self.eleValueOrTuples.pop() for _ in range(elements)]
        tupleVal = EleTuple(tupleElements)
        self.eleValueOrTuples.append(tupleVal)

    def rule10(self, tupleVal, index):
        if isinstance(tupleVal, EleTuple):
            if isinstance(index, EleValue) and index.isLabel("int"):
                ind = int(index.getValue())
                value = tupleVal.getValue()[ind]
                self.eleValueOrTuples.append(value)
                return
            raise ExceptionHandlerOfCSE("Expected integer index but found: " + index)
        raise ExceptionHandlerOfCSE("Expected tuple but found: " + tupleVal)

    def rule11(self, lambdaVal, rand):
        if isinstance(lambdaVal, EleValue) and lambdaVal.isLabel("lambda"):
            if isinstance(rand, EleTuple):
                kAndVAndC = lambdaVal.getValue().split(" ")
                k, v, c = kAndVAndC[0], kAndVAndC[1].split(","), kAndVAndC[2]
                envC = self.environments[int(c)]

                newEnvironment = Environment(envC)
                for i in range(len(v)):
                    newEnvironment.remember(v[i], rand.getValue()[i])
                newEnvIndex = str(len(self.environments))
                self.environments.append(newEnvironment)
                self.eleValues.append(EleValue("environment", newEnvIndex))
                self.eleValues.append(EleValue("delta", k))
                self.eleValueOrTuples.append(EleValue("environment", newEnvIndex))
                return
            raise ExceptionHandlerOfCSE("Expected tuple but found: " + rand)
        raise ExceptionHandlerOfCSE("Expected lambda element but found: " + lambdaVal)

    def rule12(self, lambdaVal):
        if isinstance(lambdaVal, EleValue) and lambdaVal.isLabel("lambda"):
            iAndVAndC = lambdaVal.getValue()
            etaElement = EleValue("eta", iAndVAndC)
            self.eleValueOrTuples.append(etaElement)
            return
        raise ExceptionHandlerOfCSE("Expected lambda element but found: " + lambdaVal)

    def rule13(self, gamma, eta):
        if isinstance(eta, EleValue) and eta.isLabel("eta"):
            iAndVAndC = eta.getValue()
            lambdaVal = EleValue("lambda", iAndVAndC)
            newGamma = EleValue("gamma")

            self.eleValueOrTuples.append(eta)
            self.eleValueOrTuples.append(lambdaVal)

            self.eleValues.append(gamma)
            self.eleValues.append(newGamma)
            return
        raise ExceptionHandlerOfCSE("Expected eta element but found: " + eta)
